File: iOS/PricetarPartner/PricetarPartner/Http/Requester/UpdateClientProfileRequester.swift
//
//  UpdateClientProfileRequester.swift
//  PricetarPartner
//
//  Created by Leapfrog-Inc. on 2021/02/22.
//

import UIKit

class UpdateClientProfileRequester {
        
    class func update(nickname: String, area: String, useFrequency: String, condition: String, genres: [String], options: [String], message: String, image: UIImage?, completion: @escaping ((Bool) -> ())) {
        
        var params = ["command": "updateClientProfile"]
        params["userId"] = SaveData.shared.userId
        params["nickname"] = nickname.base64Encode() ?? ""
        params["useFrequency"] = useFrequency.base64Encode() ?? ""
        params["condition"] = condition.base64Encode() ?? ""
        params["genres"] = genres.compactMap { $0.base64Encode() }.joined(separator: ",")
        params["options"] = options.compactMap { $0.base64Encode() }.joined(separator: ",")
        params["message"] = message.base64Encode() ?? ""
        
        if let image = image {
            params["image"] = image.pngData()?.base64EncodedString() ?? ""
        } else {
            params["image"] = ""
        }
        
        ApiRequester.post(params: params) { result, data in
            if result, let dic = data as? Dictionary<String, Any> {
                if (dic["result"] as? String) == "0" {
                    completion(true)
                    return
                }
            }
            completion(false)
        }
    }
}

File: Android/app/src/main/java/jp/co/lfg/pricetarpartner/Http/Requester/UpdateClientProfileRequester.java
package jp.co.lfg.pricetarpartner.Http.Requester;

import android.graphics.Bitmap;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.StringJoiner;

import jp.co.lfg.pricetarpartner.Http.ApiRequester;
import jp.co.lfg.pricetarpartner.System.Base64Utility;
import jp.co.lfg.pricetarpartner.System.BitmapUtility;
import jp.co.lfg.pricetarpartner.System.SaveData;

public class UpdateClientProfileRequester {

    public static void update(String nickanme, String area, String useFrequency, String newCondition, String oldCondition, ArrayList<String> genres, ArrayList<String> options, String message, Bitmap image, final Callback callback) {

        HashMap<String, String> params = new HashMap<>();
        params.put("command", "updateClientProfile");
        params.put("userId", SaveData.getInstance().userId);
        params.put("nickname", Base64Utility.encode(nickanme));
        params.put("area", Base64Utility.encode(area));
        params.put("useFrequency", Base64Utility.encode(useFrequency));
        params.put("newCondition", Base64Utility.encode(newCondition));
        params.put("oldCondition", Base64Utility.encode(oldCondition));

        StringJoiner genreJoiner = new StringJoiner(",");
        for (String genre : genres) {
            genreJoiner.add(Base64Utility.encode(genre));
        }
        params.put("genres", genreJoiner.toString());

        StringJoiner optionJoiner = new StringJoiner(",");
        for (String option : options) {
            optionJoiner.add(Base64Utility.encode(option));
        }
        params.put("options", optionJoiner.toString());

        params.put("message", Base64Utility.encode(message));

        if (image != null) {
            params.put("image", BitmapUtility.getBase64EncodedString(image));
        } else {
            params.put("image", "");
        }

        ApiRequester.post(params, new ApiRequester.Callback() {
            @Override
            public void didReceiveData(boolean result, JSONObject json) {
                if (result) {
                    try {
                        String apiResult = json.getString("result");
                        if (apiResult.equals("0")) {
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch (JSONException e) {}
                }
                callback.didReceiveData(false);
            }
        });
    }

    public interface Callback {
        void didReceiveData(boolean result);
    }
}
